src/common/utils/functional/functional.ts

```diff
@@ -59,23 +59,30 @@ export function threadConditionally(x: any, ...fns: Function[]) {
 }
 
 export function complement<T>(p: Predicate<T>): Predicate<T> {
- return (x: T) => !p(x);
+  return (x: T) => !p(x);
 }
 
-export function debounce<T extends (...args: any[]) => any>(fn: T, ms: number): T & { cancel: Fn } {
+// TODO: fix to use infer on arguments tuple https://stackoverflow.com/a/50014868/1089761
+export function debounceWithPromise<T extends (...args: any[]) => any>(fn: T, ms: number): ((...args: any[]) => Promise<any>) & { cancel: Fn } {
```

Comment:
Check if we could return cached promise - if we call debounced function, it should return the same promise every time. Look for inspiration in `awesome-debounce-promise`